core/src/main/java/com/alibaba/alink/operator/batch/dataproc/StratifiedSampleBatchOp.java

```diff
@@ -0,0 +1,345 @@
+import org.apache.flink.api.common.functions.GroupReduceFunction;
+import org.apache.flink.api.java.operators.GroupReduceOperator;
+import org.apache.flink.api.java.operators.UnsortedGrouping;
+import org.apache.flink.api.java.sampling.BernoulliSampler;
+import org.apache.flink.api.java.sampling.PoissonSampler;
+import org.apache.flink.api.java.sampling.RandomSampler;
+import org.apache.flink.ml.api.misc.param.Params;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.XORShiftRandom;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Random;
+
+
+public final class StratifiedSampleBatchOp extends BatchOperator<StratifiedSampleBatchOp>
+        implements SampleGroupColumnParam<StratifiedSampleBatchOp>,
+                   SampleRatioParam<StratifiedSampleBatchOp>,
+                   SampleSeedParam<StratifiedSampleBatchOp>,
+                   SampleWithReplacementParams<StratifiedSampleBatchOp> {
+
+    public StratifiedSampleBatchOp() {
+        this(new Params());
+    }
+    private StratifiedSampleBatchOp(Params params){
+        super(params.set(SEED, new Random().nextLong()));
+    }
+
+    public StratifiedSampleBatchOp(String groupKey, String ratio) {
+        this(groupKey, ratio, false);
+    }
+    public StratifiedSampleBatchOp(String groupKey, String ratio, boolean withReplacement) {
+        this(new Params()
+             .set(GROUP_COL, groupKey)
+             .set(RATIO, ratio)
+             .set(WITH_REPLACEMENT, withReplacement));
+    }
+
+    @Override
+    public StratifiedSampleBatchOp linkFrom(BatchOperator<?>... inputs) {
+        BatchOperator<?> in = checkAndGetFirst(inputs);
+        // compute index of group key
+        int index = computeGroupCloIndex(in.getColNames(), getGroupCol());
+
+        UnsortedGrouping<Row> groupingOperator = in.getDataSet().groupBy(index);
+
+        String ratio = getRatio();
+        GroupReduceOperator result = null;
+        Long seed = getSeed();
+        try {
+
+            if (compile.matcher(ratio).find()) {
+                //step2.1 if ratio is number, there is same ratio for each group, so we do nothing but sample directly
+                double r = Double.valueOf(ratio);
+                result = groupingOperator.reduceGroup(new ConsistentRatioSamplesReduce(getWithReplacement(), r, seed));
+
+            } else {
+                //step2.2 :the string ratio defines different ratio per group， we should
+                Map<Object, Double> fractionMap = new HashMap<>();
+                String[] keyRatios= ratio.split(",");
+                for (String keyRatio : keyRatios) {
+                    String[] ratios = keyRatio.split(":");
+                    fractionMap.put(ratios[0], new Double(ratios[1]));
+                }
+                result = groupingOperator.reduceGroup(new StratifiedSampleReduce(getWithReplacement(), seed, index, fractionMap));
```

Comment:
Would you mind to add a wrapper class of RandomSampler. Copy the code from flink is not a good idea if there is a workaround